### PR TITLE
[GPU] Skip redundant permute in stateful model

### PR DESCRIPTION
**Pull request by andrew-k-park**
### Details:
 - Skip permute at runtime if only one axis has valid shape and others are all 1 in the transposing axes
 - Need to check the largest value among the actual dim values in case where the permute order is different from the shape index is equal to the multiplied value

### Tickets:
 - 127443

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/permute.cpp [hunk 1]
```diff
@@ -137,7 +137,9 @@ void permute_inst::update_output_memory() {
     if (_node != nullptr)
         build_deps();
 
-    _outputs = {_network.get_engine().reinterpret_buffer(input_memory(), _impl_params->get_output_layout())};
+    GPU_DEBUG_TRACE_DETAIL << id() << " : update_output_memory with mem of input " << get_node().get_dependency(0).id()
+                           << " : " << input_memory_ptr()->buffer_ptr() << std::endl;
+    _outputs[0] = input_memory_ptr();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/permute.cpp:142]: Why not reinterpret?

Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/graph/permute.cpp]: Reverted. Previously, `update_output_memory` was also called when creating `primitive_inst`. so there was an issue of retrieving the output layout even in the case where impl_params is dynamic. I fixed this corner case itself.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/pre_dynamic_shape_opts.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+#include "pass_manager.h"
+#include "gather_inst.h"
+#include "permute_inst.h"
+#include "kv_cache_inst.h"
+#include "gemm_inst.h"
+#include "program_helpers.h"
+
+using namespace cldnn;
+
+void pre_dynamic_shape_opts::run(program& p) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/pre_dynamic_shape_opts.cpp:14]: Suggestion: How about "mark_runtime_skippable_nodes" for better understanding of  the purpose?

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 3]
```diff
@@ -981,6 +981,48 @@ void primitive_inst::do_runtime_skip_gather() {
...
 
+void primitive_inst::do_runtime_skip_permute() {
+    OV_ITT_SCOPED_TASK(ov::intel_gpu::itt::domains::intel_gpu_plugin, openvino::itt::handle("do_runtime_skip_permute: " + id()));
+    // Check pattern
+    if (!get_node().is_type<permute>()
+        || is_output()
+        || !get_node().can_be_optimized()
+        || _impl_params->has_fused_primitives()
+        || _impl_params->get_input_layout(0).data_type != _impl_params->get_output_layout().data_type
+        || (get_user_insts().size() == 1 && get_users().front()->is_type<concatenation>() && get_user_insts().front()->can_be_optimized()))
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:992]: These  conditions are already covered by get_node().can_be_optimized() so I think they are not needed any more at runtime.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/pre_dynamic_shape_opts.cpp (hunk 2 above)
Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/pre_dynamic_shape_opts.cpp]: updated as suggested

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 3 above)
Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: Removed unnecessary condition

File: src/plugins/intel_gpu/src/graph/graph_optimizer/mark_runtime_skippable_nodes.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                node.has_fused_primitives() ||
+                (impl_params->get_input_layout(0).data_type != impl_params->get_output_layout().data_type))
+                return;
+
+            // TODO: For now, all permutes with dynamic shape are applied.
+            //       A more detailed pattern will need to be applied later
+            if (node.is_dynamic()) {
+                if (node.get_dependency(0).is_type<kv_cache>())
+                    return;
+                if (node.get_dependency(0).is_type<gemm>())
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/mark_runtime_skippable_nodes.cpp:53]: Didn't we decide to allow gemm->permute pattern?

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 5]
```diff
@@ -508,8 +508,8 @@ event::ptr primitive_inst::realloc_if_needed() {
     }
 
     // Clear out memory if if was previously reused, but now primitive can't be optimized
-    if (_node->is_type<gather>() && !can_be_optimized() && _outputs[0]
-            && dep_memory_ptr(0) && _network.get_engine().is_the_same_buffer(dep_memory(0), output_memory(0))) {
+    if ((_node->is_type<gather>() || _node->is_type<permute>()) && !can_be_optimized() && _outputs[0]
+        && dep_memory_ptr(0) && _network.get_engine().is_the_same_buffer(dep_memory(0), output_memory(0))) {
         _outputs[0] = nullptr;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:513]: Also, lets let it return if it is can_be_optimized(), not allocating memory..

File: src/plugins/intel_gpu/src/graph/graph_optimizer/mark_runtime_skippable_nodes.cpp (hunk 4 above)
Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/mark_runtime_skippable_nodes.cpp]: Yes, updated again to allow for gemm's post permute pattern

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 5 above)
Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: Updated to return an empty event when can_be_optimized() is true

**Review by yeonbok — APPROVED** [review]

**Pull request merged**